app/scripts/modules/kubernetes/src/common/tolerations/TolerationsComponent.tsx

```diff
@@ -0,0 +1,210 @@
+
+  private static effects: Option[] = [
+    { value: 'NoSchedule', label: 'NoSchedule' },
+    { value: 'PreferNoSchedule', label: 'PreferNoSchedule' },
+    { value: 'NoExecute', label: 'NoExecute' },
+  ];
+
+  public tolerations: IToleration[];
+
+  constructor(props: IKubernetesTolerationsProps) {
+    super(props);
+    this.state = this.getState(props);
+  }
+
+  private getState(props: IKubernetesTolerationsProps): IKubernetesTolerationsState {
+    return {
+      tolerations: props.tolerations || [],
+    };
+  }
+
+  public componentWillReceiveProps(nextProps: IKubernetesTolerationsProps): void {
+    this.setState(this.getState(nextProps));
+  }
+
+  private addTolerations(): void {
+    const tolerations = cloneDeep(this.state.tolerations);
+    tolerations.push({ key: '', value: '', operator: 'Exists', effect: 'NoSchedule' });
+    this.props.onTolerationChange(tolerations);
+  }
+
+  private removeToleration(index: number): () => void {
+    return () => {
+      const tolerations = cloneDeep(this.state.tolerations);
+      tolerations.splice(index, 1);
+      this.props.onTolerationChange(tolerations);
+    };
+  }
+
+  private handleKeyChange(index: number): (event: any) => void {
+    return (event: any) => {
+      const tolerations = cloneDeep(this.state.tolerations);
+      tolerations[index].key = event.target.value;
+      this.props.onTolerationChange(tolerations);
+    };
+  }
+
+  private handleValueChange(index: number): (event: any) => void {
+    return (event: any) => {
+      const tolerations = cloneDeep(this.state.tolerations);
+      tolerations[index].value = event.target.value;
+      this.props.onTolerationChange(tolerations);
+    };
+  }
+
+  private handleTolerationSecondsChange(index: number): (event: any) => void {
+    return (event: any) => {
+      const tolerations = cloneDeep(this.state.tolerations);
+      tolerations[index].tolerationSeconds = event.target.value;
+      this.props.onTolerationChange(tolerations);
+    };
+  }
+
+  private handleOperatorSelect(index: number): (option: Option) => void {
+    return (option: Option) => {
+      const tolerations = cloneDeep(this.state.tolerations);
+      tolerations[index].operator = option.value as IOperator;
+      this.props.onTolerationChange(tolerations);
+    };
+  }
+
+  private handleEffectSelect(index: number): (option: Option) => void {
+    return (option: Option) => {
+      const tolerations = cloneDeep(this.state.tolerations);
+      tolerations[index].effect = option.value as IEffect;
+      this.props.onTolerationChange(tolerations);
+    };
+  }
+
+  public render() {
+    return (
+      <div>
+        <div className="sm-label-left">Tolerations</div>
+        <table className="table table-condensed">
+          <tbody>
+            {this.state.tolerations.map((row, i) => {
+              return (
+                <tr key={i}>
```

Comment:
Using an array index as a key has never bitten me, but it is supposedly bad practice. You should try to use the tolerations's `key` if it exists.

Reply:
So, default to `key` where non-empty and fall back to index when it is?

Reply:
Yep, exactly.

Reply:
I attempted to use this here but ran into some weird behavior when updating the `key` property.